### ISSUE TITLE
RTSP over websocket

### ISSUE THREAD

**Issue by stalniy** [rroller/dahua#366]
**Is your feature request related to a problem? Please describe.**

dahua implemented rtsp over websocket what makes it possible to see live video without delays 

**Describe the solution you'd like**

I was able to integrate extracted solution https://github.com/TeamDotworld/dahua-rtsp-web into home assistant for my cameras with custom hass card on UI. 

Now I need to add a WS proxy on HA side, so the connection is done from server not from browser to hide username/password

I can share/contribute this code into repository.  

Having this base, I think  it’s possible to integrate other interesting things, like draw IVS lines on camera.

**Comment by stalniy** [comment]
For anyone interested, I improved the original  project -> https://github.com/TeamDotworld/dahua-rtsp-web
added audio and IVS rendering support. Encapsulated this into webcomponent. Demo and webcomponent can be found here ->  https://github.com/stalniy/dahua-rtspoverws

I also plan to create a viewer for dahua IVS events (SMD, IVS) similar to frigate's one

**Comment by zios0007** [comment]
@stalniy that is a genuinely impressive piece of work, and the honest answer is that it is a different shape of thing from what this integration is.

What you have built is a transport and a frontend card. This integration is a Home Assistant *integration*: it talks to the device's API and produces entities, and it hands streaming to Home Assistant's own stream component and to go2rtc. Adding a websocket proxy and a custom card would mean owning a second streaming path alongside the one Home Assistant already maintains, which is a large ongoing commitment rather than a feature.

That is not a judgement on the approach. Low latency viewing is the single most common complaint about camera streams in Home Assistant, and RTSP over websocket avoids the transcoding that causes it.

The more likely route to it helping people here is go2rtc, which Home Assistant already ships and which is where stream sources are added. A source implemented there reaches every Home Assistant user rather than only users of this integration.

Leaving open as a real idea, with the link, because anyone wanting low latency Dahua video should find your project.

**Comment by stalniy** [comment]
Yes, I tried go2rtc together with Frigate. The main issue for me was that go2rtc uses CPU resources to convert RTSP streams to WebRTC.

I was running Home Assistant on an old laptop, and I didn’t want to upgrade the hardware just for the sake of running go2rtc, especially since my cameras could already deliver low-latency video directly over WebSocket.

Dahua cameras also support quite a few useful features, such as two-way audio and rendering detection zones directly over the video stream. So, in practice, they already provide most of what I needed from Frigate and go2rtc.

At one point, I even considered building a Frigate-like UI specifically for Dahua cameras, because the UI in the original firmware is really not very user-friendly. With the rise of AI-assisted development, I’ll probably have the capacity to build this soon.

**Comment by zios0007** [comment]
That is a fair objection and it changes the picture, so it is worth being accurate about where the cost actually is.

go2rtc does not transcode when it does not have to. For an H.264 stream to a browser that speaks WebRTC, it repackages rather than re-encodes, and the CPU cost is small. Where it does burn CPU is exactly where you would expect: H.265 to a client that cannot play it, or anything going out as HLS or MJPEG. If you were seeing high CPU with Frigate in the picture, Frigate's own decoding for detection is a separate and much larger cost than go2rtc's.

So "go2rtc uses CPU to convert" is true for some paths and not for others, and which path you are on is usually decided by the camera's codec rather than by go2rtc.

None of which makes your project pointless. RTSP over websocket straight from the device to the browser skips the server entirely, and that is a genuinely different trade: no transcoding, no repackaging, no server in the path at all. That is worth more on a small host than on a large one, which may be exactly your case.

The reason I would still not put it in this integration is unchanged: it would mean owning a second streaming path alongside the one Home Assistant maintains, for every user, indefinitely. As a card plus a proxy it stays optional for the people who want it, which is what you have built.

Worth knowing that setting the camera's sub stream to H.264 removes the transcoding cost from the go2rtc path for most people, since H.265 is the usual reason it appears.
